### DISCUSSION TITLE
Vercel Functions - req.url query string parameter order being re-written

### DISCUSSION THREAD

**Discussion by mynameiskyleok** [vercel/community#500] [Help]
### Observation
When using Vercel functions, the `url` property on incoming requests is being re-written, sometimes with differently ordered query string parameters than the original url. The middleware helper `query` object seems to be the in same order of the re-written `url` parameters.

### Example
Original url routed to function: `/api/user?id=123456789&email=user@domain.com`
Vercel function log of `req.url`: `/api/user?email=user@domain.com&id=123456789`

### Question
How can I get the originally URL (e.g., similar to `req.originalUrl` provided in Express) of the request? Note: It appears to be logged correctly at the top of the Vercel function logs.

### Why it Matters
I'm working with a 3rd party web services that calls the `api/user` endpoint of my Vercel hosted app. The 3rd party service sends a signature in the request header so that my Vercel hosted app can verify the origin and contents of the incoming request. Within my `api/user` endpoint logic I generate a hash to compare against the incoming request signature and in this process, the order of the query string parameters on `req.url` matters (i.e., `...?id=...&email=...` is not the same as `...?email=...&id=...` so my signature check fails.

**Comment by wiiiimm** [comment]
When you calculate the hash, instead of relying on creating a hash for the search query section of the request URL, you should parse them into different variables and reconstruct them into the order required.

**Reply by mynameiskyleok** [reply]
Thanks for the response @williamli. Unfortunately, I don't know the proper order of the parameters before the request is made to reconstruct them as you suggest. I should note that the source of the request is a commercial Sales CRM and Marketing platform and that I don't have control over the order of the parameters when the request is made. 

Is there a way to prevent the `req.url` rewrite by Vercel before executing the function? `req.url` is logged by the system correctly at the beginning of the function execution but then incorrect as soon as used in the function.

Example:
```
// Vercel system log at beginning of function execution
[GET] /api/time?id=...&email=...

// console.log(req.url)
/api/time?email=...&id=...
```
I'm only using two parameters in these examples (i.e., `id=` and `email=`) to simplify the question but in practice it could be any subset of 17 parameters.

**Reply by wiiiimm** [reply]
@mynameiskyleok I have created a [reproduction](https://github.com/vercel-support/discussion-500-nextjs-query-string-order-test/blob/main/README.md) of what you described and escalated the issue to our engineers. I will get back to you as soon as I have more info.

**Reply by erinosher** [reply]
Hey @williamli is there any news on this bug? It's also causing me some headaches with signature verification from a 3rd party service. I also found this issue on the NextJS repo about it: https://github.com/vercel/next.js/issues/35983

**Comment by someguy9** [comment]
Having the same issue! has anyone figured out a work around using a different library?
